man/ncidsip.8
.\" %W% %G%
.TH NCIDSIP 8
.SH NAME
.B ncidsip
- Inject CID info by snooping SIP invites
.SH SYNOPSIS
.B ncidsip\^
[options]
.PP
.I Options:
.PP
.nf
[-D                 | --debug]
[-d <filename>      | --dumpfile <filename>]
[-i <interface>     | --interface <interface>]
[-h                 | --help]
[-l                 | --list]
[-n <host[:port]>   | --ncid <host[:port]>]
[-s <[host][:port]> | --sip <[host][:port]>]
[-p <filename>      | --pidfile <filename>]
[-t                 | --test]
[-u                 | --usage]
[-V                 | --version]
[-v                 | --verbose]
.fi
.SH DESCRIPTION
Snoops SIP Invites via libpcap and injects the caller id information
found to the NCID server specified.  Uses Net::Pcap to interface with
the libpcap library and snoops only udp traffic on the specified SIP
host and port.
.PP
The configuration file for \fIncidsip\fR is /etc/ncid/ncidsip.conf.
.SH "OPTIONS"
.TP
.B -s <[host][:port]> | --sip=<[host][:port]>
Specifies the hostname of the SIP device to snoop.  You may also specify
the UDP port by suffixing the hostname with :<port>, or if no hostname
is wanted, just :<port>.  If you do not specify a host, it defaults to
the network interface.  If you do not specify a port, it defaults to
<5061> (Vonage default).  Other Vonage ports are 5060 and 10000.  The
new Vonage default appears to be <10000>.
.TP
.B -n <host[:port]> | --ncid=<host[:port]>
Specifies the NCID server to connect to.  Port may be specified by
suffixing the hostname with :<port>.  By default it will connect to
port 3333 on "localhost".
.TP
.B -i <interface> | --interface=<interface>
Specifies the network interface to snoop on.  If this is not specified
then libpcap will pick a network interface.  This will generally be
the first Ethernet interface found.
.TP
.B -p <filename> | --pidfile=<filename>
Specifies the pidfile name to write.
Set to \fI/var/run/ncidsip.pid\fR in a rc or init script when used as a service.
The program will still run if it does not have permission to write a pidfile.
There is no default, if pidfile is not set, no pid file will be used.
.TP
.B -h | --help
Prints this help
.TP
.B -D | --debug
Display the payload of all packets that matched the libpcap filter.
.TP
.B -d <filename> | --dumpfile <filename>
Read packets from a libpcap capture file instead of the network.
Mostly only useful for development purposes.
.TP
.B -u | --usage
Prints this help
.TP
.B -V | --version
Displays the version
.TP
.B -v | --verbose
Displays status.  Use this option to run interactive.
.TP
.B -l | --listdevs
Returns a list of all network device names that can be used.
.TP
.B -t | --test
Test for SIP packets.  This option is used to check if SIP packets
exist without starting the NCID server.  It will display the Caller
ID line generated when a call comes in, and a CANCEL line if cancel
was generated.
.SH FILES
/etc/ncid/ncidsip.conf
.SH CID MESSAGE FORMAT
.nf
CID: ###DATEmmddhhss...LINEidentifier...NMBRnumber...NAMEwords+++
.fi
.SS Example
.nf
CID: ###DATE05311233...LINE1122...NMBR13215551212...NAMEBig John+++
.fi
.SH EXAMPLES
.nf
# run ncidsip as root and view status:
ncidsip -v

# run ncidsip as root and list all network device names:
ncidsip --listdevs

# run ncidsip as root and view SIP packets:
ncidsip --debug

# run sip from the init file at boot:
ncidsip
.fi
.SH SEE ALSO
ncidsip.conf.5, sip2ncid.8, ncidd.8, ncid2ncid.1, yac2ncid.1,
ncidmodules.1, ncidrotate.1


doc/CHANGES
ncid - Network CallerID

Version 0.81
    - Makefile: updated
    - README: updated
    - ncidd.h: changed RINGWAIT from 25 to 29
               changed "CIDOUT: " to "OUT: ", added "HUP: "
                   added ttyfd and hangup to extern int
                   added extern struct termios ntty;
               added ignore1 as a extern
               added the IGNORE1 and INCLUDE1 defines
               moved to server/
    - ncidd.c: changed ringcount variable to ringwait
               changed gencid to initialize to 0 instead of 1
               added LogMsg() to display lastring and ring at level 4
               main(): removed code to print aliases
               userAlias(): added if (ignore1 && *nmbr == '1') ++nmbr;
               added hangup option
               moved to server/
    - ncidd.alias: moved to server/
    - nciddalias.c: doAlias(): changed buf[] to msgbuf[]
                               added LEVEL8 printout of alias[]
                    moved to server/
    - nciddalias.h: updated
                    moved to server/
    - nciddconf.c: added hangup to setword[] structure
                   added ignore1 to setword[] structure
                   moved to server/
    - nciddconf.h: added extern int errorStatus;
                   moved to server/
    - nciddhangup.c: added
                     moved to server/
    - nciddhangup.h: added
                     moved to server/
    - ncidd.conf-in: added hangup configuration and comments
                     added ignore1 configuration and comments
                     moved to server/
    - ncid.sh: changed CIDOUT to OUT and changed CIDOUTLOG to OUTLOG
               getCID{}: - changed if {$type < 3 || $type == 8}
                           to if {($type < 3) || ($type == 8 && $CallOut)}
                         - added: global CallOut
    -          moved to client/
    - ncid.conf-in: moved to client/
    - ncid.gif: moved to client/
    - nciddalias.c: doAlias(): added code to print aliases
                               renamed buf to msgbuf
    - ncidd.conf-in: updated for new gencid initialization
                     added ttyACM0 comment
    - moved server files to server/
    - moved client files to client/
    - cidgate/Makefile: updated
    - cidgate/ncid2ncid.c: socketConnect(): changed connect message
                           added strerror() to error messgage line 1298
                           sigdetect(): signals logged for all signals
    - debian/Makefile: updated
    - debian/README.Debian: updated
    - debian/control: updated
    - debian/ncid-hangup.init: removed
    - debian/postrm-client: updated
    - debian/prerm-client: updated
    - debian/rules: updated
    - doc/TODO: updated
    - doc/NCID-SDK.odt: updated
    - doc/Verbose-ncidd: updated
    - Fedora/Makefile: updated
    - Fedora/README.Fedora: updated
    - Fedora/ncid.spec: updated
    - Fedora/ncidsip.init: added comment
    - Fedora/ncid-hangup.init: removed
    - FreeBSD/Makefile: updated
    - FreeBSD/README.FreeBSD: updated
    - FreeBSD/ncidsip.rc: added $opts to args=... line
    - FreeBSD/ncid-hangup.rc: removed
    - man/Makefile: updated
    - man/*: updated all man pages and added man pages
    - modules/Makefile: updated
    - modules/README.modules: updated
    - modules/ncidmodules.conf-in: updated
    - modules/ncid-hangup.sh: removed
    - modules/ncid.blacklist: removed
    - modules/ncid.minicom: removed
    - server/*: added
    - test/README.test: updated
    - test/testserver: added the -B blacklist option
    - test/TestGWcalls.data: updated
    - test/testclient.data: updated
    - TiVo/Makefile: updated
    - TiVo/README.TiVo: updated
    - tools/cidcall.pl: updated and improved output
    - tools/cidupdate.pl: changed the -a and -c options to -A and -C, removed -l

Version 0.80
    - README: updated for new features
    - Makefile: added OSDCLIENT to tivo-mips: and tivo-s1:
    - ncidd.c: added global "int sendout", changed some comments in formatCID()
    - ncidd.h: added "sendout" to "extern int"
    - nciddconf.c: added "cidout" to sendclient[]
    - ncid.sh: added country codes: "NONE" for no number format,
               DE (Germany), and UK (Ubited Kingdom)
               added shell code to check NoGUI in ncid.conf at startup
               modified the ithe beginning shell code that used GUI
               added CIDOUT: CIDOUTLOG: to checkType {}
               added variable "CallOut" to display outgoing calls
               added $cidtype to the cid list in formatCID {}
               changed option -C and replaced --call-prog with --call-out
               added $cidtype to displayLog {} and displayCID {}
               added CIDOUT and CIDOUTLOG to getCID {}
               removed --call-prog from $OPTSTIVO
               removed variable "CallProg" or replaced it with Program
               changed 'set Program [list $ProgDir/$ProgName' to: set Program ""
    - ncid.conf-in: updated country code comments
               removed "set Program line and comments
               changed EXTPROG name to ProgName
    - ncidd.conf-in: corrected comment about the cid logfile
    - cidgate/sip2ncid.c: added type to the CALL: and CALLINFO lines
                          type is either CALLOUT or CALLIN
    - cidgate/sip2ncid.h: removed CIDCALL and added CALLOUT, CALLIN
                          added CALL field to CIDCAN, CIDBYE, and CIDLINE
    - debian/changelog: updated
    - debian/control: changed Requires and added Recommends for ncid-client
    - doc/NCID-SDK.odt: Updated for new features
    - Fedora/ncid.spec: updated
    - man/ncidd.8: updated for new features
    - man/ncidd.conf.8: updated for new features
    - test/TestGWcalls.data: updated to include CALLIN and CALLOUT
    - TiVo/Makefile: added OSDCLIENT
    - Win/ncid.nsi: updated

Version 0.79
    - Makefile: fixed FreeBSD problems by adding: MAKE = gmake
                CFLAGS changed -O to -O2

cidgate/ncidsip.pl
#!/usr/bin/perl

# Copyright (c) 2006 by Clayton O'Neill
# Copyright (c) 2006, 2007, 2008, 2009
# by John L. Chmielewski <jlc@users.sourceforge.net>

# ncidsip is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# any later version.

# ncidsip is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.

# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA

use strict;
use warnings;

use Net::Pcap;
use POSIX qw(strftime);
use IO::Socket::INET;
use Data::Dumper;
use Getopt::Long qw(:config no_ignore_case_always);
use Pod::Usage;

our $VERSION = "(NCID) XxXxX";

my ($host, $port) = ('localhost', 3333);
my ($siphost, $sipport) = ('', 10000);
my $dumpfile;
my $interface;
my $debug = 0;
my $verbose = '';
my ($help, $version);
my $sock = undef;
my $test = 0;
my @locals;
my $listdevs = 0;
my $pidfile = "";
my $savepid;
my $pid = 0;
my @callids;
my $callid;
my @tcallids;
my $tcallid;

my $result = GetOptions("ncid=s" => \$host,
               "dumpfile|d=s" => \$dumpfile,
               "interface=s" => \$interface,
               "sip=s" => \$siphost,
               "debug|D" => \$debug,
               "help" => \$help,
               "usage" => \$help,
               "verbose|v" => \$verbose,
               "version|V" => \$version,
               "test" => \$test,
               "listdevs" => \$listdevs,
               "pidfile|p=s" => \$pidfile
             ) || pod2usage(2);

die "ncidsip $VERSION\n" if $version;

pod2usage(-verbose => 2) if $help;

$port = $1 if $host =~ s/:(\d+)//;
$sipport = $1 if $siphost =~ s/:(\d+)//;

if ($listdevs) {
  my $err = 0;
  my @devs;
  my %devinfo;
  @devs = Net::Pcap::findalldevs(\%devinfo, \$err);
  for my $dev (@devs) {
    print "$dev : $devinfo{$dev}\n";
  }
  exit 0;
}

if ($test) {
  $verbose = 1;
  $debug = 1;
} else {
  # Only create a PID file if $pidfile is set to one.
  if ($pidfile ne "") {
    if (-e $pidfile) {
      # only one instance per computer permitted
      unless (open(PIDFILE, $pidfile)) {
        die "pidfile exists and is unreadable: $pidfile\n";
      }
      $savepid = <PIDFILE>;
      close(PIDFILE);
      chop $savepid;
      if (-d "/proc/$savepid") {
        die "Process ($savepid) already running: $pidfile\n";
      } else {
        print "Found stale pidfile: $pidfile\n" if $verbose;
      }
    }

    if (open(PIDFILE, ">$pidfile")) {
      print(PIDFILE "$$\n");
      $pid = $$;
      close(PIDFILE);
      print "Wrote pid $pid in $pidfile\n" if $verbose;
    }else{
      print "Could not write pidfile: $pidfile\n" if $verbose;
    }
  }

  &connectNCID;
  errorExit("Could not connect to NCID server: $!\n") if !defined $sock;
}

my ($err, $pcap);
if (defined $dumpfile) {
  $pcap = Net::Pcap::open_offline($dumpfile, \$err);
  errorExit("Could not open $dumpfile: $err\n") if defined $err;
} else {
  if (!defined $interface) {
    $interface = Net::Pcap::lookupdev(\$err);
      errorExit("No interface name provided and was unable to find one: $err\n")
        if defined $err;
  }
  print "Opening '$interface'\n" if $verbose;
  $pcap = Net::Pcap::open_live($interface, 1500, 1, 100, \$err);
  errorExit ("Could not open $interface: $!\n") if defined $err;

  my ($net, $mask);
  Net::Pcap::lookupnet($interface, \$net, \$mask, \$err);
  errorExit("Could not get interface information: $!\n") if defined $err;

  my $filter;
  my $filter_str = "host $siphost and port $sipport and udp";
  if (!$siphost =~ /\w/) { $filter_str =~ s/.*(port.*)/$1/; }
  print "filter: $filter_str\n" if $verbose;

  my $res = Net::Pcap::compile($pcap, \$filter, $filter_str, 0, $mask);
  $err = Net::Pcap::geterr($pcap);
  errorExit("Could not compile filter: $err\n") if $res != 0;
  Net::Pcap::setfilter($pcap, $filter);
}

$SIG{'HUP'} = 'sigHandle';
$SIG{'INT'} = 'sigHandle';
$SIG{'QUIT'} = 'sigHandle';
$SIG{'TERM'} = 'sigHandle';

Net::Pcap::loop($pcap, -1, \&processPacket, undef);

sub processPacket {
  my ($userdata, $header, $packet) = @_;
  my $ip = substr($packet, 14);
  my $iplen = (unpack('C', $ip) & 0x0F) * 4;
  my $udp = substr($ip, $iplen);
  my $sip = substr($udp, 8);
  my $name;
  my $number;
  my $tonumber;
  my $line;
  my $toline;
  my $msg;

  print $sip,"\n" if $debug;

  if (!$test) {
    if (defined $sock) {while (<$sock>) {print if $verbose;}}
    # $! (errno) == 0 if server disconnected
    # $sock undefined if could not connect to server
    if ($! == 0 || !defined $sock) {&connectNCID};
    return if !defined $sock;
  }

  my $date = strftime("%m%d%H%M", localtime($header->{tv_sec}));
  if ($sip =~ /^CSeq:\s+\d+\s+INVITE/imo) {
      # Start of Call
      if (!$sip =~ /^SIP\/.*Request Terminated/imo) {return}
      if ($sip =~ /^Call-ID:(.+)@/imo) {
        $callid = $1;
      } else {print "No INVITE Call-ID found in packet\n" if ($debug);}

      if (grep(/$callid/, @callids)) {
        # Call already processed
        return;
      } else {
        # New Call
        push(@callids, $callid);
        print "Adding $callid to CallID list\n" if ($debug);
      }

      if ($sip =~ /^To:\s+<sip:(.+)(\w\w\w\w)@/imo) {
        $tonumber = "$1$2";
        $toline = $2;
      } else {$line = "UNKNOWN";}

      if ($sip =~ /^From:\s+\"?(.+?)"?\s+<sip:/imo) {
        $name = $1;
      } else {$name = "NO NAME";}

      if ($sip =~ /^From:.*<sip:(.+)(\w\w\w\w)@/imo) {
        ($number) = "$1$2";
        $line = $2;
        } else {$number = "NO NUMBER";}

      if (grep(/^$number$/, @locals)) {
        # Outgoing Call
        $msg = sprintf("CALLINFO: ###CALLED...DATE%s...LINE%s...NMBR%s+++",
                        $date, $line, $tonumber);
      } else {
        # Incoming Call
        $msg = sprintf("CALL: ###DATE%s...LINE%s...NMBR%s...NAME%s+++",
                        $date, $toline, $number,$name);
      }

      print $msg, "\n" if $verbose;
      if (!$test) { print $sock $msg, "\r\n"; }

  } elsif ($sip =~ /^CSeq:\s+\d+\s+CANCEL/imo) {
      # Hangup before answer
      if ($sip =~ /^Call-ID:(.+)@/imo) {
        $callid = $1;
      } else {print "No CANCEL Call-ID found in packet\n" if ($debug);}
      @tcallids = @callids;
      @callids = ();
      foreach $tcallid (@tcallids) {
        if ($tcallid ne $callid) {
          push (@callids, $tcallid);
        } else {
          if ($sip =~ /^From:.*<sip:(.+)(\w\w\w\w)@/imo) {
            ($number) = "$1$2";
            $line = $2;
          } else {$number = "NO NUMBER";}

          if ($sip =~ /^To:.*<sip:(.+)(\w\w\w\w)@/imo) {
            ($tonumber) = "$1$2";
            $toline = $2;
          } else {$tonumber = "NO NUMBER";}

          if (grep(/^$number$/, @locals)) {
            # number is in the To line instead of From line
            $msg = sprintf("CALLINFO: ###CANCEL...DATE%s...LINE%s...NMBR%s+++",
                            $date, $line, $tonumber);
          } else {
            $msg = sprintf("CALLINFO: ###CANCEL...DATE%s...LINE%s...NMBR%s+++",
                            $date, $toline, $number);
          }

          print "Removed $callid from CallID list\n" if ($debug);
          print $msg, "\n" if $verbose;
          if (!$test) { print $sock $msg, "\r\n"; }
        }
      }

  } elsif ($sip =~ /^CSeq:\s+\d+\s+BYE/imo) {
      # Hangup After answer
      if ($sip =~ /^Call-ID:(.+)@/imo) {
        $callid = $1;
      } else {print "No BYE Call-ID found in packet\n" if ($debug);}
      @tcallids = @callids;
      @callids = ();
      foreach $tcallid (@tcallids) {
        if ($tcallid ne $callid) {
          push (@callids, $tcallid);
        } else {
          if ($sip =~ /^From:.*<sip:(.+)(\w\w\w\w)@/imo) {
            ($number) = "$1$2";
            $line = $2;
          } else {$number = "NO NUMBER";}

          if ($sip =~ /^To:.*<sip:(.+)(\w\w\w\w)@/imo) {
            ($tonumber) = "$1$2";
            $toline = $2;
          } else {$number = "NO NUMBER";}

          if (grep(/^$number$/, @locals)) {
            # number is in the To line instead of From line
            $msg = sprintf("CALLINFO: ###BYE...DATE%s...LINE%s...NMBR%s+++",
                            $date, $line, $tonumber);
          } else {
            $msg = sprintf("CALLINFO: ###BYE...DATE%s...LINE%s...NMBR%s+++",
                            $date, $toline, $number);
          }

          print "Removed $callid from CallID list\n" if ($debug);
          print $msg, "\n" if $verbose;
          if (!$test) { print $sock $msg, "\r\n"; }
        }
      }

  } elsif ($sip =~ /^CSeq:\s+\d+\s+REGISTER/imo) {
      $sip =~ /^Contact:\s+.*?<sip:(.+?)@/imo;
      if (!grep(/^$1$/, @locals)) {
        push(@locals, $1);
        print "Adding $1 to local number list\n" if ($debug);
      }
  }
}

sub connectNCID {
  print "NCIDsip Version $VERSION\n" if $verbose;
  print "Connecting to NCID server on $host:$port\n" if $verbose;
  $sock = IO::Socket::INET->new(PeerAddr => $host,
                 PeerPort => $port,
                 Proto => 'tcp',
               );
  defined $sock || return;

  my $greeting = <$sock>;
  print "Connected: $greeting" if $verbose;

  $sock->blocking(0);
  while (<$sock>) {print if $verbose;}
}

sub sigHandle {
    my $sig = shift;
    print "Terminating: Caught SIG$sig\n" if $verbose;
    if ($pid != 0) {unlink($pidfile);}
    exit(0);
}

sub errorExit {
    printf "@_";
    if ($pid != 0)
    {
        unlink($pidfile);
        print "Removed $pidfile\n" if $verbose;
    }
    exit(-1);
}

=head1 NAME

ncidsip - Inject CID info by snooping SIP invites

=head1 SYNOPSIS

ncidsip [--sip|-s <[host][:port]>] [--ncid|-n <host[:port]>] [--debug|-D] [--dumpfile|-d <filename>] [--interface|-i <interface>] [--help|-h] [--version|-V] [--verbose|-v] [--test|-t] [--usage|-u] [--listdevs|-l] [--pidfile|p <filename>]

=head1 DESCRIPTION

Snoops SIP INVITES via libpcap and injects the caller id information
found to the NCID server specified.  Uses L<Net::Pcap> to interface
with the libpcap library and snoops only udp traffic on the specified
SIP host and port.

=head1 OPTIONS

=over 2

=item -D, --debug

Display the payload of all packets that matched the libpcap filter.

=item -d, --dumpfile <filename>

Read packets from a libpcap capture file instead of the network.
Mostly only useful for development purposes.

=item -h, --help

Prints this help

=item -i, --interface=<interface>

Specifies the network interface to snoop on.  If this is not specified
then libpcap will pick a network interface.  This will generally be
the first ethernet interface found.

=item -l, --listdevs

Returns a list of all network device names that can be used.

=item -n, --ncid=<host[:port]>

Specifies the NCID server to connect to.  Port may be specified by
suffixing the hostname with :<port>.  By default it will connect to
port C<3333> on C<localhost>.

=item -s, --sip=<[host][:port]>

Specifies the hostname of the SIP devie to snoop.  You may also
specify the UDP port by suffixing the hostname with :<port>, or
if no hostname wanted, just :<port>.  If you do not specify a host,
it defaults to the network interface.  If you do not specify a port,
it defaults to <5061> (Vonage default).

=item -p, --pidfile=<filename>

Specifies the pidfile name to write.  The program will still run if
it does not have permission to write a pidfile.
The default pidfile name is /var/run/ncidsip.pid.

=item -t, --test

Test for SIP packets.  This option is used to check if SIP packets
exist without starting the NCID server.  It will display the Caller
ID line generated when a call comes in, and a CANCEL line if cancel
was generated.

=item -u, --usage

Prints this help

=item -V, --version

Displays the version

=item -v, --verbose

Displays status.  Use this option to run interactive.

=back

=head1 REQUIREMENTS

perl 5.6 or higher

L<Net::Pcap>

=head1 BUGS

This program takes a rather ad-hoc approach to parsing UDP datagrams
to avoid additional module dependencies.  As a result, it very likely
doesn't handle VLANs correctly.  If anyone actually wants it to work
with VLANS, feel free to email me and I'll take a look at it.

=head1 AUTHOR

Clayton O'Neill <claytononeill@users.sourceforge.net>

=cut
